### PR TITLE
use "direct" write for non-partitioned python model materializations

### LINKED ISSUE #1318: [Regression] Python model build failed when Google bucket retention policy is set (starting from dbt version 1.7)
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Starting dbt version 1.7, when bucket retention policy is set, python model build will throw error Unhandled error while executing target/run/restream_bi/models/python/python_model_test.py[0m Google Cloud Dataproc Agent reports job failure. 

### Expected Behavior

In dbt version1.6, the python model does not get a build error even when when bucket retention policy is set.

### Steps To Reproduce

1. Change dbt version to 1.7 or later in Dev env
2. Set a value to bucket retention policy in bucket details in Google Cloud
3. Run `dbt run -s <python model>`

### Relevant log output

```shell
Google Cloud Dataproc Agent reports job failure.

Using the default container image
Waiting for container log creation
PYSPARK_PYTHON=/opt/dataproc/conda/bin/python
JAVA_HOME=/usr/lib/jvm/temurin-11-jdk-amd64
SPARK_EXTRA_CLASSPATH=
:: loading settings :: file = /etc/spark/conf/ivysettings.xml
/usr/lib/spark/python/lib/pyspark.zip/pyspark/pandas/__init__.py:49: UserWarning: 'PYARROW_IGNORE_TIMEZONE' environment variable was not set. It is required to set this environment variable to '1' in both driver and executor sides if you use pyarrow>=2.0.0. pandas-on-Spark will set it for you but it does not work if there is a Spark context already launched.
```

### Environment

```markdown
- OS:macOS
- Python/dbt: 
Error in dbt cloud version 1.7 and versionless
OK in dbt cloud version 1.6
```

### Which database adapter are you using with dbt?

bigquery

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by colin-rogers-dbt**
resolves #1318

In order to support partitioned materializations in BQ python models we we switched from "direct" to "indirect" mode when writing model results in Dataproc back to BigQuery. As the naming implies "indirect" temporarily stages data in the provided GCS bucket. If a user has a retention policy on the bucket this will fail as the bucket won't allow Dataproc to delete these temp files as it goes. 

This PR sidesteps that issue to ensure backwards compatibility with <1.7 created models by using "direct" write when a partitioned config is not provided.  

Note: I have not added any new testing to cover the case where a user has set a retention policy. Ultimately I think this is an edge case we don't need to test against but we should document that a bucket retention policy cannot be used with a partitioned python model as a follow up

[docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) dbt-labs/docs.getdbt.com/#

### Checklist

- [ ] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: dbt/include/bigquery/macros/materializations/table.sql [hunk 1]
```diff
@@ -113,10 +113,19 @@ else:
   msg = f"{type(df)} is not a supported type for dbt Python materialization"
   raise Exception(msg)
 
+# For writeMethod we need to use "indirect" if materializing a partitioned table
+# otherwise we can use "direct". Note that indirect will fail if the GCS bucket has a retention policy set on it.
+{%- if partition_config is not none %}
```
Comment (VersusFacit) [inline dbt/include/bigquery/macros/materializations/table.sql:118]: nit: I'd flip this because it just reads easier to have the user not balance a negative in their head
